### PR TITLE
Support certificate-validity that populates the notAfter field in NewOrder

### PR DESCRIPTION
**Pull request by aaomidi**
## Pull Request Checklist

- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Add or update any documentation as needed to support the changes in this PR.
- [x] Include your name in `AUTHORS.md` if you like.
- [x] Add test cases for the new feature.
  - [x] Unit test for acme
  - [x] Unit test for certbot
  - [x] Integration test for certbot
- [ ] Renewals? Storage of correct metadata?

### CODE REVIEW

**Comment by aaomidi** [comment]
This is ready for review. Please let me know if you want any wording changes as I just came up with them on the spot. Please also let me know what, if anything, I should do for unit/integration testing this.

**Review by alexzorin — CHANGES_REQUESTED** [review]
Thanks, looks good. You wrote this really quickly!

Other than my inline comments,

We will need to store the expiry delta value against each certificate lineage as well (in `/etc/letsencrypt/renewal/*.conf`). Otherwise, it will be forgotten at the next renewal. To do this, you can look at how `--preferred-chain`/`preferred_chain` is implemented.

We also require some tests:

1. Something in `acme.ClientV2Test.test_new_order` to ensure the `not_after` makes it into the order resource.
2. Something in `certbot.ClientTest` to ensure that the CLI relative seconds value is propagated through `obtain_certificate*` into an absolute `not_after` in`acme.new_order`.
3. An integration test in `certbot-ci/certbot_integration_tests/certbot_tests/test_main.py`, to check that the feature works end-to-end and that the resulting certificate has the expected expiry date. Pebble has `notBefore`/`notAfter` suppport, you can use `pytest.skip` to skip Boulder, as some other tests do.

File: certbot/certbot/_internal/cli/__init__.py
Comment (alexzorin) [reply certbot/certbot/_internal/cli/__init__.py:375]: Fair enough. We can sit on it for a while and gather some other opinions.

Maybe `--certificate-lifetime`? But I am deathly paranoid that the other shoe will drop and some other CA will implement `notBefore` support, which will be an even bigger naming problem.

**Review by alexzorin — COMMENTED** [review]
If we're going to be adding support for different certificate lifetimes, we're probably also going to need to do something about `renew_before_expiry`, or renewal won't work as expected. It currently defaults to `30 days` and can only be overridden by manually modifying the renewal parameters, after the fact.

Potentially:

- If not otherwise specified, we can modify it to automatically calculate a suitable duration based on the lifetime of the certificate we received. Let's Encrypt's advice is "We recommend renewing certificates automatically when they have a third of their total lifetime left", but that needs to be validated for different lifetimes I think.
- We might want to add `--renew-before-expiry` as a CLI flag.

I think we'd have to do this as a separate piece of work, don't worry about it for now.

**Comment by alexzorin** [comment]
FYI that macOS coverage test is flaky, you should be able to hit re-run on it from https://github.com/certbot/certbot/pull/9253/checks until it passes.

**Comment by aaomidi** [comment]
> We will need to store the expiry delta value against each certificate lineage as well

I did see the expiration of the cert is being stored there - alongside what you said above with renewals being a separate piece of work - do you think anything else is left with this PR?

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (aaomidi) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:910]: FWIW I did test this test with a new version of pebble I manually built and it passed.

Comment (alexzorin) [reply certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:910]: Although this doesn't work locally on macOS, we do not want to skip this test in CI, which will run against Linux and Windows targets.

To make the integration tests that use Pebble succeed on macOS, I can share the workaround I use. Build Pebble locally on your Mac, and then copy the built Pebble binary to `./certbot-ci/certbot_integration_tests/assets/pebble_v2.3.0_linux-amd64`.  You should then be able to run `tox -e integration` successfully (assuming you have `nginx` installed as well).

Comment (aaomidi) [reply certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:910]: The problem is that pebble doesn't have a release cut with notAfter implemented.

From what I saw in the code, during the CI testing it uses the releases from GH, right? Might be imagining things.

**Comment by alexzorin** [comment]
>I did see the expiration of the cert is being stored there - alongside what you said above with renewals being a separate piece of work - do you think anything else is left with this PR?

Yes, my earlier comment is still relevant:

>We will need to store the expiry delta value against each certificate lineage as well (in /etc/letsencrypt/renewal/*.conf). Otherwise, it will be forgotten at the next renewal. To do this, you can look at how --preferred-chain/preferred_chain is implemented.

As this PR stands now, when you issue a certificate with `--certificate-validity 86400`, it works as expected.

If you then do a `certbot renew --force-renewal`, it goes back to the default validity of 90 days.

What we want to achieve is for each future renewal to remember the value of `--certificate-validity`, on a *per-certificate basis*. 

How this can be done is to persist a `certificate_validity` option to the `/etc/letsencrypt/renewal/*.conf` file, so that it gets reconstituted (restored) during renewal of each certificate.

So how do we get Certbot to store `--certificate-validity` in the renewal parameters? Take a look here:

https://github.com/certbot/certbot/blob/fe0b637e4d5acf3d24ba0d5e3feca440c2a32717/certbot/certbot/_internal/renewal.py#L42-L52

Overall, you can search around for `--preferred-chain` and `preferred_chain` (or maybe `--rsa-key-size`/`rsa_key_size` since it is also an `int`), and see how Certbot interacts with those options in order to achieve persistence of the option.

I'm going to hold off on reviewing the other parts for now, but thank you for writing all the tests!

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [reply certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:910]: >The problem is that pebble doesn't have a release cut with notAfter implemented.

Are you sure? According to https://github.com/letsencrypt/pebble/commit/2c26e5bac888b5418c53dbea203fbdf9d7516104, the feature is present in release v2.3.1.

Comment (alexzorin) [reply certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:910]: Ah, I see. Certbot downloads Pebble v2.3.0, it's hardcoded in `pebble_artifacts.py`.

If you want to open a PR to update this file, I will approve and merge it, and we can unblock your work that way.

Comment (alexzorin) [reply certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:910]: We've merged #9260 for the Pebble thing so you should be able to merge master and unskip the test.

**Comment by alexzorin** [comment]
Thanks for the work you've been putting in, I'll try to review again this very soon!

File: acme/tests/client_test.py
Comment (alexzorin) [inline acme/tests/client_test.py:761]: nit: if we're not using these in other unit tests, let's fold it all into `test_new_order_with_not_after`.

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:905]: Are these caused by an autoformatter?

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:913]: In these tests, I think it's best to generate a value using `context.get_domain(...)`, and then pass that for both `--cert-name` and `-d`, and below for `cert_path` etc as well.

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:909]:
```suggestion
    """Test that --certificate-validity is requesting certificates of the appropriate validity"""
```

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:922]: 😄 . I think it's probably worth creating our own version of `assertAlmostEqual` in `assertions.py` and using that instead. We can go with a simple implementation that just works on `int`.

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:924]: This function has a `force_renew=True` kwarg, so we can omit `--force-renewal`.

```suggestion
    context.certbot(['renew'])
```

File: certbot/certbot/_internal/cli/__init__.py
Comment (alexzorin) [reply certbot/certbot/_internal/cli/__init__.py:382]: I am still not sure what to do about the name, but I'd like for the name to contain the time unit, in accordance with the convention on (most) of our other similar flags: `--certificate-validity-seconds`.

I am also a little tempted to make it `--certificate-validity-days` and create `non_negative_float` to suit for it, but I'm on the fence with that as well. Having to type out 2592000 comes off as not so great UX to me, but I think it might require a little thought about what to do here.

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:943]: What is the logic of having a fallback date here?

If we expect `not_after_raw` to always have value, we could do something like `assert not_after_raw is not None` and let the test crash.

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:933]: Let's move this to `misc.py` along with `get_certificate`.

File: certbot-ci/certbot_integration_tests/utils/misc.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/utils/misc.py:282]:
```suggestion


```

File: certbot-ci/certbot_integration_tests/utils/misc.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/utils/misc.py:296]:
```suggestion


```

File: certbot-ci/certbot_integration_tests/utils/misc.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/utils/misc.py:288]: Just some minor nits about this docstring.

```suggestion
    Load the certificate from the provided path.
    :param str cert_path: the path to the certificate in PEM format
    :returns: the certificate
```

File: certbot/docs/cli-help.txt
Comment (alexzorin) [inline certbot/docs/cli-help.txt:1]: I assume editor autoformatting did this? Let's avoid touching this file.

File: certbot-ci/certbot_integration_tests/certbot_tests/test_main.py
Comment (alexzorin) [inline certbot-ci/certbot_integration_tests/certbot_tests/test_main.py:908]: There is one more thing I would like this test to do, and that is after the certificate issued, and also after it is renewed, to check that `certificate_validity` is present in the renewal parameters.

You can use `assert_saved_lineage_option` to do this.

**Review by alexzorin — CHANGES_REQUESTED** [review]
Thank you. 

Sorry for the delays reviewing this. To be upfront, I think we definitely need #9261 to be resolved before we can realistically merge this PR, and I'm not *yet* sure how many releases that will take. If it takes a while and you're not available to keep working on it, I will try to maintain it and eventually merge it, preserving your authorship.

Otherwise, it's looking really good. I still need to solicit some feedback about the flag name and unit, but here is another review on everything so far.

**Pull request closed without merging**
